File: src/modules/controllers/BookingController.js
class BookingController {
    constructor(bookingModel, seatController) {
        this.bookingModel = bookingModel;
        this.seatController = seatController;
    }

    generateBookingReference() {
        return 'BK-' + Math.random().toString(36).substring(2, 8).toUpperCase();
    }

    validateBooking(userId, flightId, seatNumbers) {
        const uniqueSeats = new Set(seatNumbers);
        if (uniqueSeats.size !== seatNumbers.length) {
            return { success: false, message: "Duplicate seats selected." };
        }

        if (this.bookingModel.isDuplicate(userId, flightId, seatNumbers)) {
            return { success: false, message: "Duplicate booking detected." };
        }

        return { success: true };
    }

    createBooking(userId, flightId, seatNumbers) {
        const validation = this.validateBooking(userId, flightId, seatNumbers);
        if (!validation.success) return validation;

        for (let seat of seatNumbers) {
            const result = this.seatController.selectSeat(userId, seat);
            if (!result.success) return result;
        }

        seatNumbers.forEach(seat => this.seatController.bookSeat(seat));

        const booking = {
            userId,
            flightId,
            seats: seatNumbers,
            bookingReference: this.generateBookingReference(),
            createdAt: new Date()
        };

        this.bookingModel.addBooking(booking);

        return { success: true, booking };
    }
}

module.exports = BookingController;

File: src/modules/Booking&Seat Module/controller/BookingController.java

package controller;

import model.*;
import java.util.*;

public class BookingController {

    private BookingModel bookingModel;
    private SeatController seatController;

    private PaymentService paymentService = new PaymentService();
    private FareService fareService = new FareService();
    private TicketService ticketService = new TicketService();
    private NotificationService notificationService = new NotificationService();

    public BookingController(BookingModel bookingModel, SeatController seatController){
        this.bookingModel = bookingModel;
        this.seatController = seatController;
    }

    public String generateBookingReference(){
        return "BK-" + UUID.randomUUID().toString().substring(0,6).toUpperCase();
    }

    public Booking createBooking(String userId, String flightId, List<String> seats){

        if(new HashSet<>(seats).size() != seats.size()) return null;
        if(bookingModel.isDuplicate(userId, flightId, seats)) return null;

        for(String s: seats){
            if(!seatController.selectSeat(userId, s).contains("locked"))
                return null;
        }

        double fare = fareService.calculateFare(seats.size());
        System.out.println("Fare: " + fare);

        if(!paymentService.processPayment(fare)){
            System.out.println("Payment failed");
            return null;
        }

        for(String s: seats){
            seatController.bookSeat(s);
        }

        String ref = generateBookingReference();
        Booking b = new Booking(userId, flightId, seats, ref);
        b.paymentStatus = "Paid";
        b.eTicket = ticketService.generateTicket(ref);

        bookingModel.addBooking(b);
        notificationService.sendConfirmation(userId, ref);

        return b;
    }
}
